components/brave_wallet/browser/internal/hd_key_ed25519.cc

```diff
@@ -62,74 +90,56 @@ std::string HDKeyEd25519::GetPath() const {
+  if (index >= kHardenedOffset) {
     return nullptr;
   }
+  std::array<uint8_t, 37> hmac_payload = {};
 
-  auto child_private_key = private_key_->unwrap().derive(path);
-  if (!child_private_key->is_ok()) {
-    VLOG(0) << std::string(child_private_key->error_message());
-    return nullptr;
-  }
+  auto span_writer = base::SpanWriter(base::span(hmac_payload));
+  // https://github.com/satoshilabs/slips/blob/master/slip-0010.md#private-parent-key--private-child-key
+  span_writer.Write(base::U8ToBigEndian(0));
+  span_writer.Write(private_key_);
+  span_writer.Write(base::U32ToBigEndian(index + kHardenedOffset));
+  DCHECK_EQ(span_writer.remaining(), 0u);
 
-  return std::make_unique<HDKeyEd25519>(path, std::move(child_private_key));
+  return DeriveFromHmacPayload(chain_code_, hmac_payload);
 }
 
-std::vector<uint8_t> HDKeyEd25519::Sign(base::span<const uint8_t> msg) {
-  auto signature_result = private_key_->unwrap().sign(
-      rust::Slice<const uint8_t>{msg.data(), msg.size()});
-  if (!signature_result->is_ok()) {
-    VLOG(0) << std::string(signature_result->error_message());
-    return std::vector<uint8_t>();
-  }
-  auto signature_bytes = signature_result->unwrap().to_bytes();
-  return std::vector<uint8_t>(signature_bytes.begin(), signature_bytes.end());
+// static
+std::unique_ptr<HDKeyEd25519> HDKeyEd25519::GenerateFromSeed(
+    base::span<const uint8_t> seed) {
+  return DeriveFromHmacPayload(base::byte_span_from_cstring(kMasterSecret),
+                               seed);
 }
 
-bool HDKeyEd25519::VerifyForTesting(base::span<const uint8_t> msg,
-                                    base::span<const uint8_t> sig) {
-  auto verification_result = private_key_->unwrap().verify(
-      rust::Slice<const uint8_t>{msg.data(), msg.size()},
-      rust::Slice<const uint8_t>{sig.data(), sig.size()});
-  if (!verification_result->is_ok()) {
-    VLOG(0) << std::string(verification_result->error_message());
-    return false;
+std::vector<uint8_t> HDKeyEd25519::Sign(base::span<const uint8_t> msg) {
+  // For the sake of performance `ED25519_sign` does not derive public key from
+  // private key, but assumes they are both already available as key_pair.
+  std::array<uint8_t, ED25519_PRIVATE_KEY_LEN> key_pair =
+      MakeKeyPair(private_key_, public_key_);
+
+  std::vector<uint8_t> signature(ED25519_SIGNATURE_LEN);
+  if (!ED25519_sign(signature.data(), msg.data(), msg.size(),
+                    key_pair.data())) {
+    return {};
   }
-  return true;
+  return signature;
 }
 
 std::vector<uint8_t> HDKeyEd25519::GetPrivateKeyBytes() const {
-  auto secret_key = private_key_->unwrap().secret_key_raw();
-  return {secret_key.begin(), secret_key.end()};
+  return base::ToVector(private_key_);
 }
 
 std::vector<uint8_t> HDKeyEd25519::GetPublicKeyBytes() const {
-  auto public_key = private_key_->unwrap().public_key_raw();
-  return {public_key.begin(), public_key.end()};
+  return base::ToVector(public_key_);
 }
```

Comment:
You can do this in a follow up if you prefer, but every time someone calls `GetPublicKeyBytes`, it results on a copy, as we have to wrap the value on a `vector`. This doesn't seem to be necessary and it looks like we could just have:

```
const std::array<uint8_t, kEd25519PrivateKeySize> private_key() const { return private_key_; }
```

Reply:
That in the first version of PR(actually returning span on const), but we later decided not change signature of public methods at all.

Reply:
Would you mind elaborating why?

Reply:
https://github.com/brave/brave-core/pull/26770#discussion_r1893274547

Reply:
https://github.com/brave/brave-core/pull/26770#discussion_r1893276352